src/app/code/community/Omise/Gateway/Controller/Base.php

```diff
@@ -1,19 +1,194 @@
      */
     protected function _getOrder()
     {
         $id = $this->getRequest()->getParam('order_id') ? $this->getRequest()->getParam('order_id') : null;
 
         return Mage::getModel('omise_gateway/order')->getOrder($id);
     }
+
+    /**
+     * @return Mage_Core_Controller_Varien_Action|void
+     * @throws Mage_Core_Exception
+     */
+    protected function validate() {
+        if(!isset($this->order))
+            $this->order = $this->_getOrder();
+        if (! $payment = $this->order->getPayment()) {
+            Mage::getSingleton('core/session')->addError(
+                $this->__($this->getTitle().' validation failed, we cannot retrieve your payment information. 
+                Please contact our support team to confirm the payment.'
+                )
+            );
+            $this->_redirect('checkout/cart');
+            return;
+        }
+
+        $charge = Mage::getModel('omise_gateway/api_charge')->find(
+            $payment->getMethodInstance()->getInfoInstance()->getAdditionalInformation('omise_charge_id')
+        );
+
+        if ($charge instanceof Omise_Gateway_Model_Api_Error) {
+            Mage::getSingleton('core/session')->addError($charge->getMessage());
+            return $this->_redirect('checkout/cart');
+        }
+        $this->transId = $payment->getLastTransId();
+        return $this->checkPaymentStatusAndUpdateOrder($charge, $this->order);
+    }
+
+    /**
+     * Checks payment status from $charge and updates order status accordingly.
+     * @param Omise_Gateway_Model_Api_Charge $charge
+     * @param Omise_Gateway_Model_Order $order
+     * @return Mage_Core_Controller_Varien_Action
+     */
+    protected function checkPaymentStatusAndUpdateOrder($charge, $order) {
+        if ($charge->isAwaitPayment()) {
+            return $this->paymentAwaiting($order);
+        }
+        if ($charge->isSuccessful()) {
+            return $this->paymentSuccessful($order);
+        }
+        $order->markAsFailed(
+            $this->transId,
+            $this->__('The payment was invalid, %s (%s)', $charge->failure_message, $charge->failure_code)
+        );
+        return $this->_redirect('checkout/cart');
+    }
+
+    /**
+     * If payment is awaiting to capture then updating order status as 'payment_review'. In case of 3-D secured payment,
+     * it will be 'processing'.
+     * @param Omise_Gateway_Model_Order $order
+     * @return Mage_Core_Controller_Varien_Action
+     */
+    protected function paymentAwaiting($order) {
+        $order->markAsAwaitPayment(
+            $this->transId,
+            $this->getAwaitingOrderStatus(),
+            Mage::helper('omise_gateway')->__(
+                (empty($this->getMessage()))
+                    ? 'The payment is in progress.<br/>Due to the way '.$this->getTitle().' works, this might take 
+                        a few seconds or up to an hour. Please click "Accept" or "Deny" to complete the payment manually 
```

Comment:
Probably needs rewording - "Deny" is not going to complete the payment. Maybe `Please click "Accept" or "Deny" to complete the transaction process"`